### PR TITLE
Add TypeScript support

### PR DESCRIPTION
**Pull request by ChrisAntaki**
- Adds TypeScript support
- Migrates `utils/document-ready` to TypeScript

### CODE REVIEW

File: build-system/tasks/karma.conf.js [hunk 1]
```diff
@@ -26,18 +26,22 @@ module.exports = {
 
   preprocessors: {
     'src/**/*.js': ['browserify'],
+    'src/**/*.ts': ['browserify'],
```
Comment (hartmannr76) [inline build-system/tasks/karma.conf.js:29]: does this work if we consolidate it into the previous line?
```
'src/**/*.{js,ts}': ['browserify'],
```

File: package.json [hunk 2]
```diff
@@ -100,11 +100,15 @@
...
     "through2": "4.0.2",
     "touch": "3.1.0",
+    "tsify": "5.0.4",
+    "typescript": "4.9.4",
     "vinyl-buffer": "1.0.1",
     "vinyl-source-stream": "2.0.0",
     "watchify": "4.0.0"
   },
   "optionalDependencies": {
+    "@rollup/plugin-commonjs": "24.0.0",
```
Comment (hartmannr76) [inline package.json:110]: oh, why are these optional?

Comment (ChrisAntaki) [reply package.json:110]: The `check` and `unit` GitHub action checks can ignore these dependencies, to speed up their Install steps

File: src/components/activities.js [hunk 3]
```diff
@@ -13,21 +13,22 @@
...
-import {INTERNAL_RUNTIME_VERSION} from '../constants';
-
 import {Constants} from '../utils/constants';
+import {INTERNAL_RUNTIME_VERSION} from '../constants';
 import {addQueryParam} from '../utils/url';
 
 const {
   ActivityIframePort: WebActivityIframePort,
   ActivityPorts: WebActivityPorts,
-} = require('web-activities/activity-ports');
```
Comment (hartmannr76) [inline src/components/activities.js:30]: Isn't this change not required with the commonjs plugin?

Comment (ChrisAntaki) [reply src/components/activities.js]: Good point, fixed!

File: build-system/tasks/karma.conf.js (hunk 1 above)
Comment (ChrisAntaki) [reply build-system/tasks/karma.conf.js]: Yep, updated ✅

**Comment by ChrisAntaki** [comment]
Note: The `e2e` test with Closure is failing, as is expected. #2780 will remove the Closure code, including its e2e test

File: src/utils/document-ready.ts [hunk 4]
```diff
@@ -14,30 +14,25 @@
...
  */
 
-/**
- * @param {!Document} doc
- * @return {string}
- */
-function getReadyState(doc) {
-  return /** @type {string} */ (doc['readyState']);
+function getReadyState(doc: Document): string {
+  return doc['readyState'];
```
Comment (fikus) [inline src/utils/document-ready.ts:18]: nit: doc.readyState

**Review by fikus — APPROVED** [review]
Please make sure to squash the 80+ commits when you merge to avoid polluting the change history.

File: src/utils/document-ready.ts (hunk 4 above)
Comment (ChrisAntaki) [reply src/utils/document-ready.ts]: Oh yes, that's cleaner 🧼

File: src/utils/document-ready.ts [hunk 5]
```diff
@@ -71,7 +70,7 @@ function onDocumentState(doc, condition, callback) {
  * @param {!Document} doc
  * @return {!Promise<!Document>}
```
Comment (hartmannr76) [inline src/utils/document-ready.ts:71]: Non-blocking but we should probably just aim to remove all JSDoc annotations

**Review by hartmannr76 — APPROVED** [review]

File: src/utils/document-ready.ts (hunk 5 above)
Comment (ChrisAntaki) [reply src/utils/document-ready.ts]: Good call, removed!

**Pull request merged**